// components/sidebar/index.js
import * as React from 'react';
import NavLink from './nav-link';
import Delimiter from './delimiter';
import LogoutNavItem from './logout-nav-item';

export default React.createClass({
  displayName: 'Sidebar',

  propTypes: {
    petitionSearches: React.PropTypes.arrayOf(
      React.PropTypes.shape({
        id: React.PropTypes.string.isRequired,
        label: React.PropTypes.string.isRequired,
        state: React.PropTypes.arrayOf(React.PropTypes.string).isRequired
      })
    ).isRequired
  },

  render() {
    const { petitionSearches } = this.props;
    return (
      <div className="sidebar">
        <ul style={{ margin: 0, padding: 0, listStyle: 'none' }}>
          <NavLink
            label="Petitions"
            to={{ pathname: '/petitions', query: { state: [] } }}
          >
            {
              petitionSearches.map((search) => {
                return (
                  <NavLink
                    key={search.id}
                    label={search.label}
                    to={{ pathname: '/petitions', query: { state: search.state } }}
                  />
                );
              })
            }
          </NavLink>
          <NavLink label="Users" to={{ pathname: '/users' }} />
          <NavLink label="Cities" to={{ pathname: '/cities' }} />
          <Delimiter />
          <LogoutNavItem />
        </ul>
      </div>
    );
  }
});
