// components/chip/index.js
import * as React from 'react';
import Icon from 'components/icon';


/**
 * Represents a chip - a small piece of information that can be deleted/closed.
 *
 * This can be used for lists of tags; <Chip /> is used in <MultiSelect />, e.g.
 * by Petition Finder.
 */
export default React.createClass({
  displayName: 'Chip',

  propTypes: {
    /**
     * Additional attributes to pass to the wrapping <span /> element.
     */
    attributes: React.PropTypes.object,

    /**
     * The text displayed on the <Chip />.
     */
    label: React.PropTypes.string,

    /**
     * The name of the chip; this is passed as parameter to callbacks.
     */
    name: React.PropTypes.string,

    /**
     * Callback; called when the user clicks the "x" icon.
     */
    onDelete: React.PropTypes.func.isRequired,

    /**
     * The value represented by this <Chip />.
     */
    value: React.PropTypes.string,
  },

  handleDelete(/*e*/) {
    const { onDelete, name, value } = this.props;
    onDelete(name, value, true);
  },

  render() {
    const { attributes, label } = this.props;
    return (
      <span className="chip" {...attributes}>
        <span>{label}</span>
        <span className="chip__delete" onClick={this.handleDelete}>
          <Icon icon="remove-circle" />
        </span>
      </span>
    );
  }
});
