// components/thumbnail/index.js
import * as React from 'react';

/**
 * Represents a thumbnail of an image.
 */
export default React.createClass({
  displayName: 'Thumbnail',

  propTypes: {
    /**
     * The size of the thumbnail. This assumes the thumbnail to be a square.
     * If you want to create thumbnails with an other aspect ratio than 1:1
     * then use the style prop to specify width and/or height of the thumbnail.
     */
    size: React.PropTypes.number,

    /**
     * The source url of the image to display.
     */
    src: React.PropTypes.string,

    /**
     * Additional style settings.
     */
    style: React.PropTypes.object,
  },

  getDefaultProps: () => ({
    size: 180,
  }),

  render() {
    const { size, src, style } = this.props;
    return (
      <div style={Object.assign({}, {
        width: size,
        height: size,
        backgroundSize: 'contain',
        backgroundImage: src && `url(${src})` || void 0,
        backgroundPosition: 'center center',
        backgroundColor: 'rgb(243, 243, 243)',
        backgroundRepeat: 'no-repeat',
      }, style)} />
    );
  }
});
