// components/og-data/index.js
import * as React from 'react';
import { translate } from 'config/strings';

/**
 * Displays Open Graph Data as a table.
 */
export default React.createClass({
  displayName: 'OpenGraphData',

  propTypes: {
    description: React.PropTypes.string,
    image_data: React.PropTypes.object,
    labelWidth: React.PropTypes.number,
    showAllRows: React.PropTypes.bool,
    style: React.PropTypes.object,
    title: React.PropTypes.string,
    url: React.PropTypes.string,
  },

  getDefaultProps: () => ({
    labelWidth: 100,
    showAllRows: false,
  }),

  renderRow(stringId, value, isLink = false) {
    if (value || this.props.showAllRows) {
      return (
        <tr>
          <td style={{ width: this.props.labelWidth }}>{translate(`og.${stringId}`)}</td>
          { isLink
            ? <td><a href={value} target="_blank">{value}</a></td>
            : <td>{value}</td> }
        </tr>
      );
    }
    return null;
  },

  render() {
    const { description, image_data, title, url } = this.props;
    return (
      <table ref={(c) => this._table = c} className="table table-bordered" style={this.props.style}>
        <tbody>
          {this.renderRow('url', url, true)}
          {this.renderRow('title', title)}
          {this.renderRow('description', description)}
          {this.renderRow('image-data', JSON.stringify(image_data || {}))}
        </tbody>
      </table>
    );
  }
});
