// petition/components/location-item.js
import * as React from 'react';
import Button from 'components/button';
import OpenGraphData from 'components/og-data';
import Thumbnail from 'components/thumbnail';

/**
 * Represents one list item. Shows Open Graph Data and allows hiding, deleting,
 * and refetching the location.
 */
export default React.createClass({
  displayName: 'LocationItem',

  propTypes: {
    handleAction: React.PropTypes.func.isRequired,
    item: React.PropTypes.shape({
      url: React.PropTypes.string.isRequired,
      image: React.PropTypes.string,
      description: React.PropTypes.string,
      title: React.PropTypes.string,
    }).isRequired,
  },

  /**
   * Make the thumbnail the same size as the open graph table.
   */
  componentDidMount() {
    if (this._og && this._og._table) {
      if (this._og._table.clientHeight !== this._thumbnailSize) {
        this._thumbnailSize = this._og._table.clientHeight;
        this.forceUpdate();
      }
    }
  },

  handleAction(action) {
    this.props.handleAction(action, this.props.item);
  },

  render() {
    const { item } = this.props;
    const thumbnailSize = this._thumbnailSize || 180;

    const style = {
      opacity: item.visible ? 1.0 : 0.35,
    };

    return (
      <div className="col-xs-12" style={{ marginBottom: 25 }}>
        <Thumbnail
          size={thumbnailSize}
          src={item.image}
          style={Object.assign({}, { float: 'left' }, style)}
        />
        <div style={{ marginLeft: thumbnailSize + 15 }}>
          <OpenGraphData
            ref={(c) => this._og = c}
            showAllRows
            {...item}
            style={style}
          />
          <div style={{ marginTop: -5 }}>
            <span style={{ marginRight: 10 }}>
              <Button
                name="toggle-visibility"
                onAction={this.handleAction}
                label={item.visible ? 'Hide' : 'Show'}
                />
            </span>
            <span style={{ marginRight: 10 }}>
              <Button
                name="delete"
                variation="danger"
                onAction={this.handleAction}
                label="Delete"
                />
            </span>
            <span style={{ marginRight: 10 }}>
              <Button
                name="refetch-og"
                onAction={this.handleAction}
                label="Re-Fetch OG"
                />
            </span>
          </div>
        </div>
      </div>
    );
  }
});
