// petition/components/user-info.js
import * as React from 'react';
import BasicUserInfo from 'user/components/basic-user-info';
import UserAPI from 'user/api';
import propTypes from './prop-types';

const PetitionShape = {
  owner: React.PropTypes.shape({
    id: React.PropTypes.number,
  }).isRequired,
};


/**
 * Displays basic information about the creator of the petition and allows
 * to open the user in the User Editor.
 */
export default React.createClass({
  displayName: 'UserInfo',

  propTypes: Object.assign({}, propTypes(PetitionShape)),

  getInitialState: () => ({
    user: null,
  }),

  componentWillMount() {
    this.fetchUser(this.props.petition.owner.id);
  },

  componentWillReceiveProps(nextProps) {
    const ownerId = this.props.petition.owner.id;
    const nextOwnerId = nextProps.petition.owner.id;
    if (nextOwnerId !== ownerId) {
      this.fetchUser(nextOwnerId);
    }
  },

  fetchUser(id) {
    if (id) {
      UserAPI.get(id).then(result => {
        this.setState({ user: result.data });
      }).catch(error => {
        // fail silently, we just do not show the user box, the rest of the
        // petition editor stays intact.
        // eslint-disable-next-line no-console
        console.warn(error);
      });
    }
  },

  render() {
    const { user } = this.state;
    if (!user) {
      return null;
    }
    return (
      <BasicUserInfo user={user} title="User" allowOpen />
    );
  }
});
