// petition/components/user-info.js
import * as React from 'react';
import BasicUserInfo from 'user/components/basic-user-info';
import propTypes from './prop-types';
import Card from 'components/card';

const PetitionShape = {
  owner: React.PropTypes.shape({
    id: React.PropTypes.string,
  }).isRequired,
};

const isSessionUser = (user) => {
  return user && user.id && user.id.startsWith('iris-session');
};

/**
 * Displays basic information about the creator of the petition and allows
 * to open the user in the User Editor.
 */
export default React.createClass({
  displayName: 'UserInfo',

  propTypes: Object.assign({}, propTypes(PetitionShape)),

  getInitialState: () => ({
    user: null,
  }),

  render() {
    const { owner } = this.props.petition;
    if (!owner) {
      return null;
    }
    return (
      <Card title="Owner">
        <BasicUserInfo
          user={owner}
          title="Petition Data"
          allowOpen={!isSessionUser(owner)}
          />
        { owner.data
          ? <BasicUserInfo user={owner.data} title="User" allowOpen />
          : null
        }
      </Card>
    );
  }
});
