// pages/petition-finder/index.js
/* eslint-disable react/prop-types, react/no-multi-comp */
import * as React from 'react';
import { withRouter } from 'react-router';
import makePage from 'components/page';
import Finder from 'components/finder';
import QueryStateContainer from 'helpers/query-state-container';
import tapEvent from 'helpers/tap-event';
import RoutingConstants from 'config/routing-constants';

const AutoLoadFinder = tapEvent('scroll', '#page-content-wrapper', Finder);


/* eslint-disable no-console */
function getData(skip, batchSize, config) {
  console.log(`loading data with config`, config);
  return new Promise((resolve/*, reject*/) => {
    let data = [];
    for (let i = skip; i < Math.min(135, skip + batchSize); i++) {
      data.push({
        title: `${config.sort} -- ${config.title}`,
        id: i,
        connected_locations: [],
        dc: {},
        description: '',
        images: [],
        links: [],
        owner: null,
        response_token: null,
        signatures: {
          amount: Math.floor(Math.random() * 35),
          required: 35
        },
        city: config.city || ['Aarau', 'Dornbirn', 'Zürich', 'Luzern'][Math.floor(Math.random() * 4)],
        tags: [],
        suggested_solution: '',
        type: '',
        state: ['active', 'closed', 'draft'][Math.floor(Math.random() * 3)]
      });
    }
    setTimeout(() => {
      resolve({ data, totalHits: 135 });
    }, 1000);
  });
}

/* eslint-disable no-unused-vars */
const openPetitionEditor = (id, router) => {
  console.log(`open petition editor for petition #${id}`);
};

const columns = [
  { label: 'Petition' },
  { label: 'City' },
  { label: 'Owner' },
  { label: 'Supporters' },
  { label: 'Percentage' },
  { label: 'Status' }
];

const calculateCityPercentage = ({ amount, required }) => {
  return (amount / required * 100).toFixed(2);
};

const ItemsTemplate = ({ item, router }) => (
  <tr onClick={() => openPetitionEditor(item.id, router)}>
    <td>{item.title}</td>
    <td>{item.city}</td>
    <td>{item.owner}</td>
    <td>{item.signatures.amount}</td>
    <td className="text-right">
      {calculateCityPercentage(item.signatures) || 'Unknown'}
    </td>
    <td>{item.state}</td>
  </tr>
);

export default withRouter(makePage(React.createClass({
  displayName: 'PetitionFinder',

  render() {
    const { location, router } = this.props;
    return (
      <div className="container-fluid">
        <h1>Petition Finder</h1>
        <AutoLoadFinder
          columns={columns}
          fetchData={getData}
          state={QueryStateContainer(location, router, RoutingConstants.PETITIONS.pathname)}
          itemsTemplate={ItemsTemplate}
          itemsTemplateProps={{ router }}
          searchConfig={[
            {
              name: 'title',
              id: 'titleSearchField',
              addon: 'search',
              type: 'string',
              lg: 3,
              md: 6,
              placeholder: 'Title'
            },
            {
              name: 'city',
              id: 'citySearchField',
              addon: 'search',
              type: 'autocomplete',
              lg: 3,
              md: 6,
              placeholder: 'City',
              autoCompleteStyle: {
                borderRadius: '0 4px 4px 0',
              },
              items: [
                { value: 'aarau', label: 'Aarau' },
                { value: 'aargau', label: 'Aargau' },
                { value: 'altach', label: 'Altach' },
                { value: 'alberschwende', label: 'Alberschwende' },
                { value: 'athen', label: 'Athen' }
              ],
            },
            {
              name: 'status',
              id: 'statusField',
              type: 'multi-select',
              options: [
                { value: 'signable', label: 'Signable' },
                { value: 'active', label: 'Active' },
                { value: 'draft', label: 'Draft' },
                { value: 'winner', label: 'Winner' },
                { value: 'loser', label: 'Loser' },
                { value: 'sendLetter', label: 'Send Letter' },
                { value: 'waitForLetterResponse', label: 'Wait for Letter Response' },
                { value: 'letterResponseArrived', label: 'Letter Response Arrived' },
                { value: 'closed', label: 'Closed' }
              ],
              lg: 6,
            },
            {
              xs: 6,
              name: 'clear-filters',
              label: 'Clear Filters',
              role: 'clear-filters',
            },
            {
              name: 'sort',
              id: 'sortDropdown',
              type: 'options',
              xs: 6,
              pull: 'right',
              options: [
                { value: '', label: 'Sort' },
                { value: 'date', label: 'Date' },
                { value: 'trending', label: 'Trending' },
                { value: 'amount-supporters', label: 'Amount Supporters' }
              ],
            }
          ]}
          uniqueItemProperty="id"
        />
      </div>
    );
  }
})));
