// components/button/index.js
import * as React from 'react';


export default React.createClass({
  displayName: 'Button',

  propTypes: {
    attributes: React.PropTypes.object,
    label: React.PropTypes.string,
    name: React.PropTypes.string.isRequired,
    onAction: React.PropTypes.func.isRequired,
    /**
     * The variation prop directly maps to bootstrap css classes.
     */
    variation: React.PropTypes.oneOf([
      'default',
      'danger',
      'primary',
      'success',
      'info',
      'warning',
      'link'
    ]),
  },

  getDefaultProps: () => ({
    variation: 'default',
  }),

  handleClick(/*e*/) {
    const { name, onAction } = this.props;
    onAction(name);
  },

  render() {
    const { attributes, label, name, variation } = this.props;

    return (
      <button
        aria-label={label}
        className={`btn btn-${variation}`}
        name={name}
        type="button"
        onClick={this.handleClick}
        {...attributes}
      >
        {label}
      </button>
    );
  }
});
