// user/components/basic-user-info.js
import * as React from 'react';
import { withRouter } from 'react-router';
import Button from 'components/button';
import Card from 'components/card';


/**
 * Displays basic information about a user.
 */
export default withRouter(React.createClass({
  displayName: 'BasicUserInfo',

  propTypes: {
    /**
     * Whether the component should display a button to open this user in the
     * user editor.
     */
    allowOpen: React.PropTypes.bool,

    router: React.PropTypes.shape({
      push: React.PropTypes.func.isRequired,
    }).isRequired,

    /**
     * The title of this block.
     */
    title: React.PropTypes.string,

    user: React.PropTypes.shape({
      email: React.PropTypes.string,
      firstname: React.PropTypes.string,
      id: React.PropTypes.string.isRequired,
      lastname: React.PropTypes.string,
      mobile: React.PropTypes.string,
      zip: React.PropTypes.string,
      town: React.PropTypes.string,
      street: React.PropTypes.string,
      salutation: React.PropTypes.string,
    }).isRequired,
  },

  getDefaultProps: () => ({
    allowOpen: false,
    title: 'Basic User Info',
  }),

  renderInUserEditorButton() {
    const { allowOpen, router, user } = this.props;
    if (!allowOpen) {
      return null;
    }
    const openInUserEditor = () => {
      router.push(`user/${user.id}`);
    };
    return (
      <div className="col-xs-12" style={{ marginTop: 12 }}>
        <Button
          name="open-in-user-editor"
          id="open-in-user-editor"
          onAction={openInUserEditor}
          label="Go to User"
        />
      </div>
    );
  },

  render() {
    const { title, user } = this.props;
    return (
      <Card title={title}>
        <div className="row">
          <div className="col-xs-12"><strong>Id: {user.id}</strong></div>
          <div className="col-xs-12">{user.salutation || ''} {user.firstname} {user.lastname}</div>
          { user.email
            ? <div className="col-xs-12">
                <a href={`mailto:${user.email}`}>{user.email}</a>
              </div>
            : null
          }
          { user.mobile
            ? <div className="col-xs-12">{user.mobile}</div>
            : null
          }
          { user.zip || user.town || user.street
            ? <div className="col-xs-12">
                <div>{user.zip || ''} {user.town || ''}</div>
                <div>{user.street || ''}</div>
              </div>
            : null
          }
          {this.renderInUserEditorButton()}
        </div>
      </Card>
    );
  }
}));
