// inputs/dropdown/index.js
import * as React from 'react';
import { InputPropTypes, ItemShape } from 'inputs/prop-types';


/**
 * Represents a dropdown menu that lets the user select one out of multiple
 * values.
 */
export default React.createClass({
  displayName: 'Dropdown',

  propTypes: Object.assign({}, InputPropTypes, {
    attributes: React.PropTypes.object,
    options: React.PropTypes.arrayOf(ItemShape).isRequired,
    value: React.PropTypes.string,
  }),

  getDefaultProps: () => ({
    attributes: {},
  }),

  handleChange(e) {
    const { name, onValueChange } = this.props;

    onValueChange(name, e.target.value, true);
  },

  render() {
    const { attributes, id, options, value } = this.props;
    return (
      <select
        className="form-control"
        value={value}
        id={id}
        {...attributes}
        onChange={this.handleChange}
      >
        {options.map(o => (
          <option key={o.value} value={o.value}>{o.label}</option>
        ))}
      </select>
    );
  }
});
