// petition/letter-appendix.js
import * as React from 'react';
import _ from 'lodash';
import authenticate from 'auth';
import { anonymize } from 'helpers/text';
import Alert from 'components/alert';
import PetitionApi from 'petition/api';
import SupporterApi from 'supporter/api';
import Template from './template';

export default authenticate(React.createClass({
  displayName: 'Letter-Appendix',

  propTypes: {
    params: React.PropTypes.shape({
      id: React.PropTypes.string.isRequired
    })
  },

  getInitialState() {
    return {
      error: null,
      context: {
        petition: {},
        supporters: []
      }
    };
  },

  componentWillMount() {
    this.getPetition(this.props.params.id);
  },

  getPetition(petitionId) {
    if (petitionId) {
      PetitionApi.get(petitionId, 'owner').then((response) => {
        const petition = _.get(response, 'data', {});

        this.getSupporters(petitionId, _.get(petition, 'supporters.amount', 0), 50).then((responses) => {
          const supporters = [];

          responses.forEach((res) => {
            const data = _.get(res, 'data', []);
            data.forEach((supporter) => {
              _.set(supporter, 'phone_user.telephone', anonymize(_.get(supporter, 'phone_user.telephone'), 6, 'x'));
              supporters.push(supporter);
            });
          });

          this.setState({
            error: null,
            context: {
              petition: petition,
              supporters: supporters
            }
          });
        });
      }).catch((response) => {
        this.setState({
          error: response.error && response.error.description
        });
      });
    }
  },

  getSupporters(petitionId, amount, limit) {
    const requests = [];
    for (let i = 0; i < (amount / limit); i++) {
      requests.push(
        SupporterApi.search({
          petition: petitionId,
          resolve: 'user',
          offset: i * limit,
          limit: limit
        })
      );
    }
    return Promise.all(requests);
  },

  render() {
    const { error, context } = this.state;

    if (error) {
      return (
        <div className="container-fluid">
          <Alert message={error} />
        </div>
      );
    }

    return (
      <Template className="letter letter-appendix" url="/templates/letter-appendix.html" context={context} />
    );
  }
}));
