// petition/letter-appendix.js
import * as React from 'react';
import _ from 'lodash';
import authenticate from 'auth';
import hasAdminAccess from '../auth/has-admin-access';
import Alert from 'components/alert';
import { PublicSupporterApi } from 'supporter/api';
import Template from './template';
import { translate } from 'config/strings';

export default authenticate(React.createClass({
  displayName: 'Letter-Appendix',

  propTypes: {
    location: React.PropTypes.shape({
      query: React.PropTypes.shape({
        token: React.PropTypes.string
      })
    }),
    me: React.PropTypes.shape({
      roles: React.PropTypes.arrayOf(React.PropTypes.string)
    }),
    params: React.PropTypes.shape({
      id: React.PropTypes.string.isRequired
    }),
  },

  getInitialState() {
    return {
      error: null,
      context: {
        petition: {},
        supporters: []
      }
    };
  },

  componentWillMount() {
    this.getPetition(this.props.params.id);
  },

  getPetition(petitionId) {
    if (petitionId) {
      this.getSupporters(petitionId, 50, 50).then((responses) => {
        const supporters = [];

        responses.forEach((res) => {
          const data = _.get(res, 'data', []);
          data.forEach((supporter) => {
            supporters.push(supporter);
          });
        });

        this.setState({
          error: null,
          context: {
            supporters: supporters
          }
        });
      });
    }
  },

  getSupporters(petitionId, amount, limit) {
    const isAdmin = hasAdminAccess(this.props.me);
    const token = this.props.location.query.token || null;
    let requestParams = {
      petition: petitionId,
      resolve: 'user',
      offset: 0,
      limit: limit
    };
    if (!isAdmin) {
      requestParams.token = token;
    }
    return Promise.all([
      PublicSupporterApi.search(requestParams)
    ]);
  },

  render() {
    const { error, context } = this.state;

    if (error) {
      return (
        <div className="container-fluid">
          <Alert message={error} />
        </div>
      );
    }

    const title = translate('petitions.appendix');
    return (
      <Template className="letter-appendix" title={title} url="/templates/letter-appendix.html" context={context} />
    );
  }
}), true /*allowPublicAccess*/);
