// petition/letter-cover.js
import * as React from 'react';
import _ from 'lodash';
import authenticate from 'auth';
import Alert from 'components/alert';
import { PetitionPublicAPI } from 'petition/api';
import Template from './template';
import { translate } from 'config/strings';

export default React.createClass({
  displayName: 'Letter-Cover',

  propTypes: {
    params: React.PropTypes.shape({
      id: React.PropTypes.string.isRequired
    })
  },

  getInitialState() {
    return {
      error: null,
      context: {
        petition: {}
      }
    };
  },

  componentWillMount() {
    this.getPetition(this.props.params.id);
  },

  getPetition(petitionId) {
    if (petitionId) {
      PetitionPublicAPI.get(petitionId, 'city,owner').then((response) => {
        const petition = _.get(response, 'data', {});

        this.setState({
          error: null,
          context: {
            petition: petition
          }
        });
      }).catch((response) => {
        this.setState({
          error: response.error && response.error.description
        });
      });
    }
  },

  render() {
    const { error, context } = this.state;

    if (error) {
      return (
        <div className="container-fluid">
          <Alert message={error} />
        </div>
      );
    }

    const title = translate('petitions.letter');
    return (
      <Template className="letter letter-cover" title={title} url="/templates/letter-cover.html" context={context} />
    );
  }
});
