// petition/model.js
const PetitionSupporters = (supportersObject) => ({
  count: supportersObject && supportersObject.amount || 0,
});

const getCityDisplay = (petition) => {
  if (petition && petition.city) {
    if (petition.city.data && petition.city.data.name) {
      return petition.city.data.name;
    }
    return petition.city.id;
  }
  return null;
};

const getName = (obj) => `${obj.firstname || ''} ${obj.lastname || ''}`.trim();

const getOwnerDisplay = (petition) => {
  if (petition && petition.owner) {
    const petitionUserName = getName(petition.owner);
    const userUserName = petition.owner.data && getName(petition.owner.data) || '';
    return petitionUserName || userUserName || petition.owner.id;
  }
  return null;
};

/**
 * The model of a petition.
 */
const Petition = (petitionObject) => {
  if (typeof petitionObject !== typeof {} || petitionObject === null) {
    return null;
  }
  return {
    id: petitionObject.id,
    dc: petitionObject.dc || {},
    title: petitionObject.title || 'untitled',
    city: getCityDisplay(petitionObject) || 'unknown',
    owner: getOwnerDisplay(petitionObject) || 'unknown',
    supporters: PetitionSupporters(petitionObject.supporters),
    state: petitionObject.state || 'unknown',
  };
};

export default Petition;
