// petition/model.js
const PetitionSupporters = (supportersObject) => ({
  count: supportersObject && supportersObject.amount || 0,
});

const getCityDisplay = (petition) => {
  if (petition && petition.city) {
    if (petition.city.data && petition.city.data.name) {
      return petition.city.data.name;
    }
    return petition.city.id;
  }
  return null;
};

const getOwnerDisplay = (petition) => {
  if (petition && petition.owner) {
    if (petition.owner.data) {
      const owner = petition.owner.data;
      return `${owner.firstname || ''} ${owner.lastname || ''}`;
    }
    return petition.owner.id;
  }
  return null;
};

/**
 * The model of a petition.
 */
const Petition = (petitionObject) => {
  if (typeof petitionObject !== typeof {} || petitionObject === null) {
    return null;
  }
  return {
    id: petitionObject.id,
    title: petitionObject.title || 'untitled',
    city: getCityDisplay(petitionObject) || 'unknown',
    owner: getOwnerDisplay(petitionObject) || 'unknown',
    supporters: PetitionSupporters(petitionObject.supporters),
    state: petitionObject.state || 'unknown',
  };
};

export default Petition;
