// petition/template.js
import * as React from 'react';
import nunjucks from 'nunjucks';
import Alert from 'components/alert';
import { translate } from 'config/strings';

export default React.createClass({
  displayName: 'Template',

  propTypes: {
    url: React.PropTypes.string,
    context: React.PropTypes.object
  },

  getInitialState() {
    return {
      // error: null,
      template: null
    };
  },

  componentWillMount() {
    this.loadTemplate(this.props.url);
  },

  componentWillReceiveProps(nextProps) {
    if(this.props.url !== nextProps.url) {
      this.loadTemplate(nextProps.url);
    }
  },

  loadTemplate(url) {
    const env = new nunjucks.Environment(new nunjucks.WebLoader('.', {async: true}));
    env.getTemplate(url, true, (err, tmpl) => {
      // TODO: error handling
      this.setState({
        // error: err,
        template: tmpl
      });
    });
  },

  render() {
    const { template } = this.state;

    // if(error) {
    //   return (
    //     <div className="container-fluid">
    //       <Alert message="Fehler!" />
    //     </div>
    //   );
    // }

    let html;
    if(template) {
      html = template.render(this.props.context);
    }

    return (
      <div dangerouslySetInnerHTML={{__html: html}} />
    );
  }
});
