// inputs/prop-types.js
import { PropTypes } from 'react';

/**
 * Common component properties, all input components require.
 */
export const InputPropTypes = {
  /**
   * Additional attributes that should be passed to the native element.
   */
  attributes: PropTypes.object,

  /**
   * Used as the element's id in the DOM.
   *
   * This is required to wire the input element to a <label />.
   */
  id: PropTypes.string.isRequired,

  /**
   * The name of the component as used as parameter in callbacks.
   */
  name: PropTypes.string.isRequired,

  /**
   * Called when the input component's value changes.
   *
   * The supplied parameters are:
   * - name: string; The name of the component as specified by the name prop.
   * - value: any; The new value of the input component.
   * - isValid: boolean; Whether the new value is valid or not.
   */
  onValueChange: PropTypes.func.isRequired,
};

/**
 * Shape of collection items; used by components that have to deal with multiple
 * values (i.e. <Dropdown />, <Autocomplete />).
 */
export const ItemShape = PropTypes.shape({
  /**
   * The actual value this item represents.
   */
  value: PropTypes.string.isRequired,

  /**
   * The human-readable representation of the value.
   */
  label: PropTypes.string.isRequired,
}).isRequired;
