// components/finder/search-form-utils.js
import Button from 'components/button';
import Autocomplete from 'inputs/autocomplete';
import ApiAutocomplete from 'inputs/api-autocomplete';
import CityAutocomplete from 'city/autocomplete';
import TextInput from 'inputs/text';
import Dropdown from 'inputs/dropdown';
import MultiSelect from 'inputs/multi-select';


export const enhanceSearchConfig = (searchConfig) => {
  return Object.assign({}, searchConfig, {
    addon: searchConfig.addon || '',
    defaultValue: searchConfig.defaultValue || '',
    label: searchConfig.label || '',
    showLabel: searchConfig.showLabel || false,
    lg: searchConfig.lg || searchConfig.md || searchConfig.xs || 12,
    md: searchConfig.md || searchConfig.xs || 12,
    placeholder: searchConfig.placeholder || '',
    role: searchConfig.role || '',
    type: searchConfig.type || 'string',
    value: searchConfig.value || searchConfig.defaultValue || '',
    xs: searchConfig.xs || 12,
  });
};

export const inputComponentConfigs = {
  'autocomplete': {
    component: Autocomplete,
    props: {},
  },
  'api-autocomplete': {
    component: ApiAutocomplete,
    props: {},
  },
  'city-autocomplete': {
    component: CityAutocomplete,
    props: {},
  },
  'multi-select': {
    component: MultiSelect,
    props: {},
  },
  'options': {
    component: Dropdown,
    props: {}
  },
  'string': {
    component: TextInput,
    props: {}
  },
};

export const roleComponentConfigs = {
  'clear-filters': {
    component: Button,
    props: {
      name: 'clear-filters',
      label: 'Clear Filters',
    }
  },
};
